### PR TITLE
lodash - change weakmap declaration to work with TS 2.2

### LINKED ISSUE #14324: Lodash All declarations of 'WeakMap' must have identical type parameters.
- [x] I tried using the `@types/lodash` package and had problems.
- [ ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @....

I found this problem compiling a project using Typescript version 2.2.0-dev.20170128 and @types/lodash version 4.14.51. In my case the tsconfig use es6 target. The error message is:

`node_modules/@types/lodash/index.d.ts(19421,15): error TS2428: All declarations of 'WeakMap' must have identical type parameters.
`
If I browse the file to the line indicated in the error message I can see the following comment:

`// Backward compatibility with --target es5`

Maybe this is the cause of the issue?
Best regards to all

### PR DESCRIPTION
**Pull request by budiadiono**
In TS 2.2 WeakMap interface for ES6 has been changed from WeakMap<K, V> to WeakMap<K extends object, V>. This changes makes @types/lodash not working with new ES6 introduced in TS2.2.

To keep the backward compatibility with ES5, I think it is safe to rename the WeakMap declaration.

I know TS 2.2 is still in RC version now, but many people using it, including me :). So if you think this is too early and also bad idea then feel free to close this PR without merging it.

If this get merged then it's should close #14324.

This pull request is clean version of pull request #14502

Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/Microsoft/TypeScript/pull/13350
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by TonyPythoneer** [comment]
I appreciate your idea but it's too early. It's because TypeScript hasn't released stable 2.2 version yet.
I hope this proposal can be an issue and make up a TODO list. It doesn't make you work in vain.

**Comment by corbinu** [comment]
Can this be merged now that 2.2 is stable?

**Comment by Dirrk** [comment]
Builds breaking waiting for this change

**Comment by chrootsu** [comment]
:+1:

**Comment by sshquack** [comment]
Just to echo what others have already mentioned here, this problem is causing issues with the TypeScript 2.2 release

https://github.com/Microsoft/TypeScript/issues/14256

File: lodash/index.d.ts [hunk 1]
```diff
@@ -12701,7 +12701,7 @@ declare namespace _ {
          * @param value The value to check.
          * @returns Returns true if value is correctly classified, else false.
          */
-        isWeakMap<K, V>(value?: any): value is WeakMap<K, V>;
+        isWeakMap<K, V>(value?: any): value is WeakMapLike<K, V>;
```
Comment (mhegazy) [inline lodash/index.d.ts]: this change disables the guard.  `WeakMapLike` does not even have the declarations of `WeakMap`, so it is not usable.

I would say just remove `WeakMap` declaration altogether, and make this function return boolean.

**Review by mhegazy — CHANGES_REQUESTED** [review]

File: lodash/index.d.ts (hunk 1 above)
Comment (JProgrammer) [inline lodash/index.d.ts]: @mhegazy would #14787 be a better solution?

Comment (mhegazy) [inline lodash/index.d.ts]: I do not think so. this change seems good to me, but we just need to remove the weakMap reference altogether.

Comment (budiadiono) [inline lodash/index.d.ts]: @mhegazy thanks for the review. I'll update it now.

**Comment by andy-ms** [comment]
A better fix would be to have `isWeakMap(value?: any): value is WeakMap<any, any>;`.
(It doesn't make sense to have type parameters because generics are erased, and `isWeakMap<{ x: number }, boolean>(foo)` would still return `true` even if `foo` is a `WeakMap<{ y: string }, Date>`.)

Then the test must be changed to use e.g. `WeakMap<{ x: number }, number>` because of the type constraint on WeakMaps.

**Comment by mhegazy** [comment]
> A better fix would be to have isWeakMap(value: any): value is WeakMap<any, any>;.

@andy-ms, the issue is not the user-defined type guard, the issue is the declaration of `WeakMap`. `WeakMap` is not defined in `lib.d.ts` (only `lib.es2015.collections.d.ts` or one of its dependents); so lodash had a forward declaration of it. TS 2.2 changed the declaration to include `T extends object` on the key, now lodash needs to do the same, but `object` is a TS 2.2 feature.  so you see it is a catch-22.

Ultimiatlly https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14787 is the correct fix, but this unblocks ppl at the moment.

**Review by mhegazy — APPROVED** [review]

**Pull request merged**